lib/html5ever/precompiled.ex

```diff
@@ -0,0 +1,278 @@
+defmodule Html5ever.Precompiled do
+  @moduledoc false
+
+  require Logger
+
+  @available_targets ~w(aarch64-apple-darwin x86_64-apple-darwin x86_64-unknown-linux-gnu x86_64-unknown-linux-musl arm-unknown-linux-gnueabihf x86_64-pc-windows-msvc x86_64-pc-windows-gnu)
+  @available_nif_versions ~w(2.14 2.15 2.16)
+
+  @doc """
+  Returns the target triple for download or compile and load.
+
+  This function is translating and adding more info to the system
+  architecture returned by Elixir/Erlang to one used by Rust.
+
+  The returning string format is the following:
+
+    "nif-NIF_VERSION-ARCHITECTURE-VENDOR-OS-ABI"
+
+  ## Examples
+
+      iex> target()
+      {:ok, "nif-2.16-x86_64-unknown-linux-gnu"} 
+
+      iex> target()
+      {:ok, "nif-2.15-aarch64-apple-darwin"}
+
+  """
+  def target(config \\ target_config()) do
+    sys_arch = maybe_override_with_env_vars(config.system_arch)
+
+    arch_os =
+      case config.os_type do
+        {:unix, os} ->
+          os
+          |> normalize_arch_os(sys_arch)
+          |> system_arch_to_string()
+
+        {:win32, _} ->
+          # 32 or 64 bits
+          arch =
+            case config.word_size do
+              4 -> "i686"
+              8 -> "x86_64"
+              _ -> "unknown"
+            end
+
+          sys_arch
+          |> Map.put_new(:arch, arch)
+          |> Map.put_new(:vendor, "pc")
+          |> Map.put_new(:os, "windows")
+          |> Map.put_new(:abi, "msvc")
+          |> system_arch_to_string()
+      end
+
+    cond do
+      arch_os not in @available_targets ->
+        {:error, "precompiled NIF is not available for this target: #{inspect(arch_os)}"}
+
+      config.nif_version not in @available_nif_versions ->
```

Comment:
Hrm, I am thinking a little about this. This has the following issue: if there is a new OTP version with a new nif version, then you cannot update the OTP version your app uses unless we go back in time and release new precompiled packages for all previous versions.

So I am thinking we should pick the more recent compatible NIF version:

    defp find_compatible_nif_version(vsn, available) do
      if vsn in available do
        vsn
      else
        [major | _] = String.split(".")

        available
        |> Enum.filter(fn vsn -> String.starts_with?(vsn, major <> ".") end)
        |> Enum.max_by(fn vsn -> vsn |> String.split(".") |> Enum.map(&String.to_integer/1) end)
    end

Reply:
OK, I will try that. Thanks!

Reply:
I kept the error message in case we can't find a compatible version (likely when a major change occurs).